File: tensorflow/go/data/mnist.go
// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package data

import (
	"compress/gzip"
	"encoding/binary"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"os"
)

type MnistImage []byte
type MnistLabel uint8

// DataSet stores MNIST image, label data for both training/testing.
type DataSet struct {
	pos         int
	TrainImages []MnistImage
	TrainLabels []MnistLabel
	TestImages  []MnistImage
	TestLabels  []MnistLabel
}

const (
	url            = "http://yann.lecun.com/exdb/mnist/"
	trainImageFile = "train-images-idx3-ubyte.gz"
	trainLabelFile = "train-labels-idx1-ubyte.gz"
	testImageFile  = "t10k-images-idx3-ubyte.gz"
	testLabelFile  = "t10k-labels-idx1-ubyte.gz"
	imageMagic     = 0x00000803
	labelMagic     = 0x00000801
	Width          = 28
	Height         = 28
)

func maybeDownload(dirname string, filename string) {
	filepath := fmt.Sprintf("%s/%s", dirname, filename)
	_, err := os.Stat(filepath)
	if err == nil {
		log.Println(filename, "is already downloaded")
		return
	}
	out, err := os.Create(filepath)
	if err != nil {
		log.Fatal("Fail to create:", err)
		return
	}

	log.Println("Downloading", filename)
	res, err := http.Get(url + filename)
	if err != nil {
		log.Fatal(err)
	}
	defer res.Body.Close()
	n, err := io.Copy(out, res.Body)
	if err != nil || n == 0 {
		log.Fatal("Fail to write file:", err)
		return
	}
}

func extractImages(dirname string, filename string) (images []MnistImage, err error) {
	name := fmt.Sprintf("%s/%s", dirname, filename)
	f, err := os.Open(name)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	z, err := gzip.NewReader(f)
	if err != nil {
		return nil, err
	}

	var (
		magic int32
		n     int32
		nrows int32
		ncols int32
	)

	if err = binary.Read(z, binary.BigEndian, &magic); err != nil {
		return nil, err
	}
	if magic != imageMagic {
		return nil, os.ErrInvalid
	}
	if err = binary.Read(z, binary.BigEndian, &n); err != nil {
		return nil, err
	}
	if err = binary.Read(z, binary.BigEndian, &nrows); err != nil {
		return nil, err
	}
	if err = binary.Read(z, binary.BigEndian, &ncols); err != nil {
		return nil, err
	}

	images = make([]MnistImage, n)
	size := nrows * ncols
	for i := 0; i < int(n); i++ {
		images[i] = make(MnistImage, size)
		size_, err := io.ReadFull(z, images[i])
		if err != nil {
			return nil, err
		}
		if size_ != int(size) {
			return nil, os.ErrInvalid
		}
	}

	return images, nil
}

func extractLabels(dirname string, filename string) (labels []MnistLabel, err error) {
	name := fmt.Sprintf("%s/%s", dirname, filename)
	f, err := os.Open(name)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	z, err := gzip.NewReader(f)
	if err != nil {
		return nil, err
	}
	var (
		magic int32
		n     int32
	)

	if err = binary.Read(z, binary.BigEndian, &magic); err != nil {
		return nil, err
	}
	if magic != labelMagic {
		return nil, os.ErrInvalid
	}
	if err = binary.Read(z, binary.BigEndian, &n); err != nil {
		return nil, err
	}
	labels = make([]MnistLabel, n)
	for i := 0; i < int(n); i++ {
		var l MnistLabel
		if err := binary.Read(z, binary.BigEndian, &l); err != nil {
			return nil, err
		}
		labels[i] = l
	}
	return labels, nil
}

// Create MNIST data set
//
// Arguments:
//   dirname: Directory where MNIST raw file is downloaded
//
// Returns DataSet which includes TrainImages,TrainLabels,TestImages,TestLabels
func NewDataSet(dirname string) *DataSet {
	files := []string{
		trainImageFile,
		trainLabelFile,
		testImageFile,
		testLabelFile,
	}

	for _, f := range files {
		maybeDownload(dirname, f)
	}

	trainImages, err := extractImages(dirname, trainImageFile)
	if err != nil {
		log.Fatal(err)
		return nil
	}
	trainLabels, err := extractLabels(dirname, trainLabelFile)
	if err != nil {
		log.Fatal(err)
		return nil
	}
	testImages, err := extractImages(dirname, testImageFile)
	if err != nil {
		log.Fatal(err)
		return nil
	}
	testLabels, err := extractLabels(dirname, testLabelFile)
	if err != nil {
		log.Fatal(err)
		return nil
	}

	return &DataSet{
		0,
		trainImages,
		trainLabels,
		testImages,
		testLabels,
	}
}

func (ds *DataSet) updatePos(max int) {
	ds.pos += max
}

// Returns the next training batch
//
// Arguments:
//   max: The max samples of next batch
//
func (ds *DataSet) NextBatch(max int) ([]MnistImage, []MnistLabel) {
	if len(ds.TrainImages) < ds.pos {
		// All samples were read
		return nil, nil
	}
	length := int(math.Min(float64(max), float64(ds.pos+max)))
	defer ds.updatePos(length)
	return ds.TrainImages[ds.pos:length], ds.TrainLabels[ds.pos:length]
}

File: tensorflow/contrib/kfac/examples/mnist.py
# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Utilities for loading MNIST into TensorFlow."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf

__all__ = [
    'load_mnist',
]


def load_mnist(data_dir,
               num_epochs,
               batch_size,
               flatten_images=True,
               use_fake_data=False):
  """Loads MNIST dataset into memory.

  Args:
    data_dir: string. Directory to read MNIST examples from.
    num_epochs: int. Number of passes to make over the dataset.
    batch_size: int. Number of examples per minibatch.
    flatten_images: bool. If True, [28, 28, 1]-shaped images are flattened into
      [784]-shaped vectors.
    use_fake_data: bool. If True, generate a synthetic dataset rather than
      reading MNIST in.

  Returns:
    examples: Tensor of shape [batch_size, 784] if 'flatten_images' is
      True, else [batch_size, 28, 28, 1]. Each row is one example.
      Values in [0, 1].
    labels: Tensor of shape [batch_size]. Indices of integer corresponding to
      each example. Values in {0...9}.
  """
  if use_fake_data:
    rng = np.random.RandomState(42)
    num_examples = batch_size * 4
    images = rng.rand(num_examples, 28 * 28)
    if not flatten_images:
      images = np.reshape(images, [num_examples, 28, 28, 1])
    labels = rng.randint(10, size=num_examples)
  else:
    mnist_data = tf.contrib.learn.datasets.mnist.read_data_sets(
        data_dir, reshape=flatten_images)
    num_examples = len(mnist_data.train.labels)
    images = mnist_data.train.images
    labels = mnist_data.train.labels

  dataset = tf.contrib.data.Dataset.from_tensor_slices((np.asarray(
      images, dtype=np.float32), np.asarray(labels, dtype=np.int64)))
  return (dataset.repeat(num_epochs).shuffle(num_examples).batch(batch_size)
          .make_one_shot_iterator().get_next())
